modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaAssetEntryAssetCategoriesCheck.java

```diff
@@ -0,0 +1,208 @@
+			newContent = JavaSourceUtil.addImports(
+				newContent,
+				"com.liferay.asset.entry.rel.service." +
+					"AssetEntryAssetCategoryRelLocalService",
+				"org.osgi.service.component.annotations.Reference");
+
+			newContent = StringUtil.replaceLast(
+				newContent, CharPool.CLOSE_CURLY_BRACE,
+				"\n\t@Reference\n\tprivate " +
+					"AssetEntryAssetCategoryRelLocalService\n\t\t" +
+						"_assetEntryAssetCategoryRelLocalService;\n\n}");
+		}
+
+		return newContent;
+	}
+
+	private String _replaceAddOrDeleteAssetCategories(String content) {
+		String newContent = content;
+
+		Matcher addOrDeleteAssetEntryAssetCategoriesMatcher =
+			_addOrDeleteAssetEntryAssetCategoriesPattern.matcher(content);
+
+		while (addOrDeleteAssetEntryAssetCategoriesMatcher.find()) {
+			String methodCall =
+				addOrDeleteAssetEntryAssetCategoriesMatcher.group();
+
+			if (!hasClassOrVariableName(
+					"AssetCategoryLocalService", newContent, methodCall)) {
+
+				continue;
+			}
+
+			String line = getLine(
+				content,
+				getLineNumber(
+					content,
+					addOrDeleteAssetEntryAssetCategoriesMatcher.start()));
+
+			String indent = SourceUtil.getIndent(line);
+
+			String newLine;
+			String newMethodCall;
+			String secondParameter =
+				addOrDeleteAssetEntryAssetCategoriesMatcher.group(2);
+
+			String variableTypeName = getVariableTypeName(
+				newContent, newContent, secondParameter, true);
+
+			if (variableTypeName.equals("List<AssetCategory>")) {
+				newLine = StringBundler.concat(
+					indent, "for (AssetCategory assetCategory : ",
+					secondParameter, ") {\n\t", line);
+
+				newMethodCall = StringUtil.replace(
+					methodCall, secondParameter,
+					"assetCategory.getCategoryId()");
+			}
+			else {
+				newLine = StringBundler.concat(
+					indent, "for (long assetCategoryId : ", secondParameter,
+					") {\n\t", line);
+
+				newMethodCall = StringUtil.replace(
+					methodCall, secondParameter, "assetCategoryId");
+			}
+
+			newContent = StringUtil.replaceFirst(newContent, line, newLine);
+
+			newMethodCall = StringBundler.concat(
+				newMethodCall, StringPool.SEMICOLON, StringPool.NEW_LINE,
+				indent, StringPool.CLOSE_CURLY_BRACE);
+
+			newMethodCall = StringUtil.replace(newMethodCall, "(\n", "(\n\t");
+			newMethodCall = StringUtil.replace(newMethodCall, ",\n", ",\n\t");
+
+			String methodStart =
+				addOrDeleteAssetEntryAssetCategoriesMatcher.group(1);
+
+			if (methodStart.contains("addAssetEntryAssetCategories")) {
+				newMethodCall = StringUtil.replace(
+					newMethodCall, methodStart, _NEW_ADD_METHOD);
+			}
+			else {
+				newMethodCall = StringUtil.replace(
+					newMethodCall, methodStart, _NEW_DELETE_METHOD);
+			}
+
+			newContent = StringUtil.replaceFirst(
+				newContent, methodCall + StringPool.SEMICOLON, newMethodCall);
+		}
+
+		return newContent;
+	}
+
+	private String _replaceAddOrDeleteAssetCategory(String content) {
+		String newContent = content;
+
+		Matcher addOrDeleteAssetEntryAssetCategoryMatcher =
```

Comment:
Same thing here. You can just call it `matcher`.

Reply:
done